cpp/onedal/array.hpp

```diff
@@ -32,54 +34,155 @@ class array {
   friend array<Y> const_array_cast(const array<U>&);
 
   public:
-    explicit array(std::int64_t element_count)
-        : data_(new T[element_count]), // TODO: Use custom oneDAL allocator
-          size_(element_count) {}
-
-    explicit array(const detail::shared<T>& data, std::int64_t size)
-        : data_(data),
-          size_(size) {}
+    array()
+        : data_owned_ptr_(nullptr),
+          size_(0),
+          capacity_(0) {}
+
+    explicit array(std::int64_t element_count, T element = {})
+        : data_owned_ptr_(new T[element_count], // TODO: Use custom oneDAL allocator
+                [](T* obj){ delete[] obj; }),
+          data_(data_owned_ptr_.get()),
+          size_(element_count),
+          capacity_(element_count) {
+        for (std::int64_t i = 0; i < size_; i++) {
+            (*this)[i] = element;
+        }
+    }
 
     template <typename Deleter>
     explicit array(T* data, std::int64_t size, Deleter&& deleter)
-        : data_(data, deleter),
-          size_(size) {}
+        : data_owned_ptr_(data, std::forward<Deleter>(deleter)),
+          data_(data),
+          size_(size),
+          capacity_(size) {}
 
     template <typename U>
     array(const array<U>& other)
-        : data_(other.data_),
-          size_(other.size_) {}
+        : data_owned_ptr_(other.data_owned_ptr_),
+          data_(other.data_),
+          size_(other.size_),
+          capacity_(other.capacity_) {}
+
+    T* get_mutable_data() const {
+        return std::get<T*>(data_);
+    }
+
+    const T* get_data() const {
+        if (auto ptr_val = std::get_if<T*>(&data_)) {
+            return *ptr_val;
+        } else {
+            return std::get<const T*>(data_);
+        }
+    }
 
-    T* get_pointer() const noexcept {
-        return data_.get();
+    bool has_mutable_data() const {
+        return std::holds_alternative<T*>(data_) && (get_mutable_data() != nullptr);
     }
 
-    std::int64_t get_size() const noexcept {
+    void unique() {
```

Comment:
I thought `unique` should return `*this` if reference counter == 1, otherwise return deep copy of the array. Per my understanding, it should not modify the object.

Reply:
- Even if the reference counter is equal to 1, `array` can point to user-defined memory which we cannot mutate. In this case, we should do a deep copy.
-  The option when we return new "unique" copy of array from this method and not change the state of the current object was proposed by me sometime before. We agreed on our meetings that this case of modifying current object is better.